### PR TITLE
bmclib version 2

### PR DESCRIPTION
**Pull request by joelrebel**
This version is where all  features and fixes for the newer [client](https://github.com/bmc-toolbox/bmclib/blob/master/client.go#L24) interface methods are to be implemented going ahead.

Clients are to import this version as `github.com/bmc-toolbox/bmclib/v2`.

I'm looking for suggestions on if this should be a separate `v2` branch instead of a directory. 

The older [interface](https://github.com/bmc-toolbox/bmclib/blob/master/devices/interfaces.go) methods will continue to be maintained as is.

The only difference as of now with the previous bmclib release
is that the Device object type returned by the `Inventory()` methods are
of the `github.com/bmc-toolbox/common.Device` type.

### CODE REVIEW

**Comment by jacobweinstock** [comment]
Hey @joelrebel , curious if there are some specific challenges with the current code structure this is seeking to solve?  I don't believe versioning changes with this new code structure, right?

**Comment by joelrebel** [comment]
Hey @jacobweinstock, thanks for taking a look, theres a few reasons for the suggested `v2`, since raising this PR, I realize a branch `v2` is recommended over a directory `v2`.

The thought process behind this is,

   - The  _older_ [ScanAndConnect](https://github.com/bmc-toolbox/bmclib/blob/master/discover/discover.go) and related [interfaces](https://github.com/bmc-toolbox/bmclib/blob/master/devices/interfaces.go) will stay in maintenance mode.
   - The _newer_ [client](https://github.com/bmc-toolbox/bmclib/blob/master/client.go) and related [interfaces](https://github.com/bmc-toolbox/bmclib/tree/master/bmc)  gets new hardware support and features, the `v2` release will not include the _older_  client and interface methods.
   
At the moment, theres one breaking change to the _newer_ client `Inventory()` methods, where the device object returned
is now of the `common.Device` type.

Apart from the this, 

 a.  I'd like bmclib to have one client interface that clients can use (instead of two) `v2` gives us that way ahead.
  
b. The _older_ interface methods get pretty much zero new hardware support and have limited maintenance being done. Its hardware compatibility with _older_ interface methods is questionable, from what I see, a lot of the _older_ interface methods do not work with newer Dells and Supermicros, and so I'd like to get rid of code that is not in use.

d. I'd like to also add support for newer hardware without wading through all the cruft of the older unmaintained codebase and working around conflicts in the method and variable names - which shows up when I'm adding new hardware support for the _newer_ client. 

c. In `v2` the _newer_ client methods can also be adapted to use Go Generics, reducing quite a bit of boilerplate code there.  

The choice of `v2` as the branch name - given that there is no `v1` is because Go expects a major version release that is not on the master branch to be `v2` and higher, along with the idea is that clients explicitly import `github.com/bmc-toolbox/bmclib/v2` so that they do not end up upgrading by mistake. 

bmclib should have been tagged a `v1.x.x` a while ago, but we have been conservative in the tagging and so I'm inclined to jump straight to `v2`  with these set of changes which

**Comment by jacobweinstock** [comment]
This is great! Thanks for sharing this reasoning. 

What makes you say a v2 branch is preferred over a v2 directory? From this [post]("Go Modules: v2 and Beyond - The Go Programming Language" https://go.dev/blog/v2-go-modules), I believe the v2 directory is preferred, no?

**Comment by joelrebel** [comment]
Based on what I understood from https://github.com/golang/go/wiki/Modules#releasing-modules-v2-or-higher
having the `v2.*.*` tags on a separate `v2` branch reduces confusion to some extent.

The other reason was that a `v2` directory would require separate CI, linter related configuration while they sit on the same branch. I'm still open to a `v2` directory if that makes more sense/is easier to work with in this case.

**Comment by chrisdoherty4** [comment]
FWIW, I advocate branches over subdirectories as I've found it much easier to maintain over a sub directory with respect to how Go modules work and git workflows like cherry-picking fixes.

**Review by mmlb — COMMENTED** [review]
I only made it about 1/3 of the way so far but now I'm off for a while so might as well add what I've got.

File: v2/bmc/firmware.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+			metadataLocal.SuccessfulProvider = elem.name
+			return status, metadataLocal, nil
+		}
+	}
+
+	return status, metadataLocal, multierror.Append(err, errors.New("failure in FirmwareInstallStatus"))
+}
+
+// FirmwareInstallStatusFromInterfaces pass through to library function
+func FirmwareInstallStatusFromInterfaces(ctx context.Context, installVersion, component, taskID string, generic []interface{}) (status string, metadata Metadata, err error) {
```
[the remark this answers is not in the document]
Comment (joelrebel) [reply v2/bmc/firmware.go]: The [method is invoked](https://github.com/bmc-toolbox/bmclib/blob/6cb51bb34666125f4c36232c1406a6abbcaa9528/client.go#L248) with [`GetDriverInterfaces`](https://github.com/jacobweinstock/registrar/blob/main/registrar.go#L79) which returns various interfaces that may or may not implement the `FirmwareInstallVerfier` interface.

File: v2/bmc/postcode.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	"github.com/hashicorp/go-multierror"
+	"github.com/pkg/errors"
+)
+
+// PostCodeGetter defines methods to retrieve device BIOS/UEFI POST code
+type PostCodeGetter interface {
+	// PostCode retrieves the BIOS/UEFI POST code from a device
+	//
+	// returns 'status' which is a (bmclib specific) string identifier for the POST code
+	// and 'code' with the actual POST code returned to bmclib by the device
```
[the remark this answers is not in the document]
Comment (joelrebel) [reply v2/bmc/postcode.go]: Good point, added it as an issue here https://github.com/bmc-toolbox/bmclib/issues/281

File: v2/bmc/postcode.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+			}
+			metadataLocal.SuccessfulProvider = elem.name
+			return status, code, metadataLocal, nil
+		}
+	}
+
+	return status, code, metadataLocal, multierror.Append(err, errors.New("failure to get device POST code"))
+}
+
+// GetPostCodeFromInterfaces is a pass through to library function
```
[the remark this answers is not in the document]
Comment (joelrebel) [reply v2/bmc/postcode.go]: fixed

File: v2/bmc/power.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+// powerProviders is an internal struct to correlate an implementation/provider and its name
+type powerProviders struct {
+	name             string
+	powerStateGetter PowerStateGetter
+	powerSetter      PowerSetter
+}
+
+// SetPowerState sets the power state for a BMC, trying all interface implementations passed in
+// if a metadata is passed in, it will be updated to be the name of the provider that successfully executed
+func SetPowerState(ctx context.Context, state string, p []powerProviders) (ok bool, metadata Metadata, err error) {
```
[the remark this answers is not in the document]
Comment (joelrebel) [reply v2/bmc/power.go]: thanks, fixed

**Comment by joelrebel** [comment]
Thanks for all the comments and suggestions here, they have been incorporated into the latest set of commits.
The [v2 code base](https://github.com/bmc-toolbox/bmclib/tree/v2) is now smaller and cleaner which should allow quicker addition of new features and fixes.

Since all of this code is currently functional and running in production, I'm going to leave the remaining suggestions as is, with issues added for some. 

I'll now close this PR since the base branch is `master`.  The PR branch `v2` will stay and be tagged with `v2.0.0`.

**Pull request closed without merging**